libtenzir/builtins/aggregation-functions/sum.cpp

```diff
@@ -44,7 +49,216 @@ class sum_function final : public aggregation_function {
   std::optional<type_to_data_t<Type>> sum_ = {};
 };
 
-class plugin : public virtual aggregation_function_plugin {
+class sum_instance : public aggregation_instance {
+public:
+  using sum_t = variant<int64_t, uint64_t, double, duration>;
+  sum_instance(ast::expression expr) : expr_{expr} {
+  }
+
+  auto update(const table_slice& input, session ctx) -> void override {
+    auto s = eval(expr_, input, ctx);
+    // XXX: Need to review again
+    auto f = detail::overload{
+      [](const arrow::NullArray&) {},
+      [&](const arrow::Int64Array& array) {
+        // UInt64 => UInt64
+        // Double => Double
+        if (not sum_) {
+          sum_ = int64_t{};
+        }
+        sum_ = sum_->match(
+          [&](int64_t self) -> sum_t {
+            auto overflow = false;
+            for (auto i = int64_t{}; i < array.length(); ++i) {
+              if (array.IsValid(i)) {
+                auto result = checked_add(self, array.Value(i));
+                if (not result) {
+                  overflow = true;
+                  continue;
+                }
+                self = result.value();
+              }
+            }
+            if (overflow) {
+              diagnostic::warning("integer overflow").primary(expr_).emit(ctx);
+            }
+            return self;
```

Comment:
You have to remember when overflow occurs. The return value should then be `null`.